### ISSUE TITLE
Fixing Systems in ODE.py

### ISSUE THREAD

**Issue by emath** [sympy/sympy#12276]
I have fixed many of the problems with systems in ODE.py.  See the attached notebook (its Python 3 -- huge native Jordan exp takes several pages -- useful info but definitely skip ahead): 

But I need advice, guidance, help, ...   before I start generating pull requests, which I also am not very good at.   Here are my initial thoughts. 

a) I'm concerned about speed.  In particular, I'm using **integrate** very naively and any help on how to do sympy integrals better and faster  would be welcome.  I have a "look-up" table approach in mind, but it will be labor intensive (good student project) and may not overcome issues in all cases (???). 

b) Some of the "fixes" not only repair issues, but also would require significant change.  For example, I never could get the **is_complex__** etcetera approach to work for all cases, so I modified the Jordan Block code to return essentially the same thing (complex eigs ==> sines/cosines ).  Also, it works for all linear systems, but some of the special types produce errors or different "looking" solutions, and it should be at the discretion of you all as to when my systems code is called.  

c)  Aaron sent me some initial feedback and referenced issues  #9235, #11718.  I'm not sure what to do about them -- i.e., would I need pull requests addressing those pulls, or just one massive "rewrite" 

Finally, I have developed my own tests (see notebook) as this is part of a larger project on splitting and composition methods for symplectic integrators.   Not sure if you would want other parts of this larger project -- either as a standalone or in connection with ode.py (see end of notebook _dsolveSystem).  

I will contribute one part of the project -- inverse Laplace Transforms -- if they are still needed once I finish developing them.     

Thanks and looking forward to your suggestions. 

Jeff
[ODEPyLinSys3.zip](https://github.com/sympy/sympy/files/831876/ODEPyLinSys3.zip)
[ODEPyLinSys3.pdf](https://github.com/sympy/sympy/files/831875/ODEPyLinSys3.pdf)

**Comment by oscarbenjamin** [comment]
This looks excellent. I'm sorry no one responded at the time you opened this issue. Did any of this work make it in to sympy?

**Comment by emath** [comment]
Thank you for the comment.  To my knowledge, none of it was incorporated
into sympy.

I would be happy to work on it again, but I no longer have a "research"
teaching load (Im teaching 4 classes and have three master's students
currently).

Probably would be summer before I could contribute further.

Thanks again!

Jeff

**Comment by oscarbenjamin** [comment]
I've written a roadmap for ODE systems:
https://github.com/sympy/sympy/wiki/ODE-Systems-roadmap

I'm hoping that a student will take this up as a GSOC project this summer.

It looks like someone is interested (#18720)

Are you happy for someone to use the work that you have done here? Also if you have any feedback on the roadmap that would be great.

**Comment by emath** [comment]
Absolutely!  Would be thrilled for a student to take it on!!  Huge fan of
GSOC!!!     Be glad to help in any way I can!

And I'll make time if you need me to help getting it set up as a GSOC
project.

Thanks!!

Jeff

**Comment by emath** [comment]
I would be glad to see this as a GSOC project.  And I'll make time if you
need me to help getting it set up as a GSOC project.

I glanced at the roadmap.  It looks good, but I'll take a better look when
I get the chance.  The Jordan Normal form is numerically unstable, but for
closed form solutions it is often the best because eigenvalues are exact
rather than approximate.  That given, if the Jordan form causes issues,
there are ways around it -- e.g., get schur then use row operations to
reduce to bidiagonal form (not necessarily 1's above the diagonal).
Exponential of a bidiagonal can be expressed in closed form (schur-parlett
algorithm).

But probably too much for a GSOC.  I mention only in that if a student does
get frustrated with the Jordan form due to its numerical instability --
which can occur -- then maybe just tell them to put  a hook there and I can
submit the more advanced version for the cases they are struggling with (if
any -- may not be an issue, but has been an issue in other venues and
wouldn't want them spinning their wheels too much dealing with it).

Thanks!!

Jeff

**Comment by oscarbenjamin** [comment]
The primary goal for sympy's ODE module is always closed form solutions. I can see use in e.g. being able to compute numerical solutions to high accuracy but I imagine there are easier ways to do that within sympy and I know that numerical solution of ODEs is well covered by other libraries anyway.

What I find is that the primary limitation of the Jordan normal form technique is if the eigenvalues are very complicated or can not be found in closed form. I think any analytic solution method will suffer from the same problem though because they all need the eigenvalues somewhere: they are an unavoidable part of the closed form solution.

**Comment by emath** [comment]
Yes, I agree -- I would not (and do not) use Sympy to compute numerical
solutions.  Agree absolutely that the primary limitation of Jordan is if
eigenvalues are complicated or not obtainable in closed form.   It appears
that the way Sympy matrix exponential works is to request eigenvalues and
eigenvectors for a matrix, where the eigenvalues are closed form roots of
the characteristic polynomial.  Then somewhere a decision must be made as
to whether or not an eigenvalue has multiplicity > 1, which can be really
hard to do even with closed form expressions.  Jordan blocks are quiet
rigid by definition, and that inflexibility makes the Jordan step possibly
both difficult and frustrating in calculating closed form matrix
exponentials.

But here is what I suggest we call a trick -- not something I would spring
on a student but something very useful in symplectic integrators, which are
in part "staying closed form for as long as we can" numerical algorithms.

If you know the eigenvalues and eigenvectors of an n x n matrix, then the
schur form can be constructed in n or fewer steps by conjugating with
Householder transforms that reflect eigenvectors to standard basis vectors
(which produce the 0's below the diagonal of the final form).  This is as
exact as the eigenvalues, eigenvectors are, as the Householders are finite
algebraic expressions in closed form.  As already mentioned, we can then
calculate the matrix exponential in closed form from the resulting closed
form schur.    Very cool!  Avoids the Jordan form difficulties.  However,
we would need to test it thoroughly before adopting it, because there is
always a Lie Group structure when we do it with symplectic and haven't
thought about it enough generically.

So if the GSOC person gets bogged down or frustrated with Jordan, then
there is probably a work-around for at least some of the difficulties.
JIC, I'll move the matrix exponential via "exact schur" testing to front
burner this summer and will do it with sympy-generated eigenvalues,
eigenvectors so it is ready if needed.

**Comment by oscarbenjamin** [comment]
That sounds useful!

Deciding whether the eigenvalue has multiplicity > 1 does not seem to be a problem. In the cases where sympy is able to find analytic roots at all it does not seem to have a problem with this.

I think the hard part of the Jordan form method is deciding whether a repeated eigenvalue has full geometric multiplicity which corresponds to whether it has trivial Jordan blocks. This is done by computing the null-space of `(M - lambda I)`. We need to solve this anyway though in order to get the eigenvectors because if the matrix is defective then a full set of eigenvectors does not exist.

**Comment by emath** [comment]
The method we use is this:  Given n eigenvalues, with multiplicities, of an
n times n matrix A
1. Find a unit eigenvector of A w.r.t a given eigenvalue
2. Householder conjugate with eigenvector -> [1,0,...,0]^T reflector
3. Select next eigenvalue and lower left (n-1) x (n-1) submatrix as A
4. Repeat until done

Result in n steps or fewer is an exact upper triangular schur form. Note,
we get eigenvectors one at a time from the increasingly smaller lower left
matrices -- only the first is necessarily an actual eigenvector of the
original matrix.

There are difficulties.  Sometimes Householder conjugation reduces more
than one column at a time.  Ordering of the eigenvalues may be
important???   Ill-conditioned???  When the goal is closed form
expressions, not sure what else may occur.  Will find out asap if this is
feasible in general ( = summer).  More than happy for anyone else to try it
out sooner if they so desire.

**Comment by oscarbenjamin** [comment]
A recent PR (#18797) added a bidiagonalisation algorithm. It seems that it can only work if it can determine whether `|A| <= |B|` for two vectors `A` and `B`. That places a significant limitation on its applicability in a symbolic context. Determining that `a<b` is a lot easier than determining that `a-b=0` if the expressions can be evaluated numerically with precision but where there are indeterminate symbols it is still subject to Richardson's theorem:
https://en.wikipedia.org/wiki/Richardson%27s_theorem
